### ISSUE TITLE
Fail to recognize Generics in Superclass

### ISSUE THREAD

**Issue by W3D3** [SpoonLabs/gumtree-spoon-ast-diff#96]
**Input:**
1.java
```java
package com.test;

public class TestClass extends SuperClass<Old> {

}

```
2.java
```java
package com.test;

public class TestClass extends SuperClass<New> {

}

```
**Output:**
```bash
java -jar gumtree-spoon-ast-diff-1.17-jar-with-dependencies.jar 1.java 2.java
no AST change
```

I would have expected the change "Old > New" to be represented in the AST.
This is happening because this is a `CtTypeReference` and all `CtReference`s are ignored.

**Comment by sedflix** [comment]
The current version fails to identify changes in the superclass. 
Example:
```
$ cat 1.java
package com.test;

public class TestClass extends SuperClass1 {

}

$ cat 2.java
package com.test;

public class TestClass extends SuperClass2 {

}
```

The output produced by it is: `no AST change`. It happens because of the `element.isImplicit()` check [here](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/361364ddaad3e8f0921a068275d4d21b6e836e85/src/main/java/gumtree/spoon/builder/TreeScanner.java#L78). 

It is interesting to note that the generics in the superclass is not identified even if we consider all the elements type. In other words, changing the method body of `isToIgnore` to `return false`, doesn't help in recognizing the generics.

**Comment by monperrus** [comment]
which element is implicit here?

**Comment by sedflix** [comment]
Removing the implicit check result in the following output for the input containing just different SuperClass.
```
Update Invocation at com.test.TestClass
	super() to super()
```
Therefore, `super()`  is the implicit element.

**Comment by monperrus** [comment]
is that in the default implicit constructor?

**Comment by sedflix** [comment]
Yeah, I think so. Its parent is `TestClass() {\n}`.

**Comment by monperrus** [comment]
@sedflix any news?

**Comment by sedflix** [comment]
If we want "extends" arguments to be compared separately and not invocation as shown above where the change is `super() to super()`, we need a check like below in [isToIgnore()](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/361364ddaad3e8f0921a068275d4d21b6e836e85/src/main/java/gumtree/spoon/builder/TreeScanner.java#L71)

```
        if (element instanceof CtTypeReference && element.getRoleInParent() == CtRole.SUPER_TYPE && element.getParent() instanceof CtClass) {
                return false;
        }
```
Adding the above check ensures that generics and extends arguments are present in the SpoonGumTree. Adding this is not enough, as the "Update" check is done through `label` which are currently `""` for our elements of interests. Therefore, all of our elements of interests will be equals in the [this check](https://github.com/GumTreeDiff/gumtree/blob/1439d6657bbb5ce6aad6db261fbaa57fc811b7eb/core/src/main/java/com/github/gumtreediff/actions/ActionGenerator.java#L112)

To solve the label problem, I propose to add the following method to gumtree.spoon.builder.LabelFinder:
```
    @Override
    public void scanCtReference(CtReference reference) {
        super.scanCtReference(reference);
        if(label.equals("") && reference instanceof CtTypeReference && reference.getRoleInParent() == CtRole.SUPER_TYPE && reference.getParent() instanceof CtClass){
            label = reference.toString();
        }
    }
```

Doing the above changes results in failure of following test cases:

### [1](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/dbd2720f414cf6fbd6a2c22211e33a5098fb7c07/src/test/java/gumtree/spoon/TreeTest.java#L190)
![image](https://user-images.githubusercontent.com/21156678/58032888-c9218600-7b23-11e9-8966-dbd89b54faea.png)

### [2](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/361364ddaad3e8f0921a068275d4d21b6e836e85/src/test/java/gumtree/spoon/AstComparatorTest.java#L1524)

The actions now include `Update TypeReference at BehaviorCall:5 Callback<T> to Callback`

### [3](https://github.com/SpoonLabs/gumtree-spoon-ast-diff/blob/361364ddaad3e8f0921a068275d4d21b6e836e85/src/test/java/gumtree/spoon/AstComparatorTest.java#L1685)
Similar as above.

I think there are better ways to add labels and check conditions in isToIgnore which I'm not aware of. 
Is an approach on this line fine, or do we need an alternative one?

**Comment by monperrus** [comment]
Thanks. Could you fix the test cases in #98 directly to see the actual impact (the first impact on JSON is not a problem for example).

**Comment by sedflix** [comment]
Done!
